# Analyzing the hiring success of community events at Kiwi.com

## Final project for Czechitas Digital Academy

The lines bellow are my first Python lines ever. I've been learning how to code for the past two months. I am aware that there are better, quicker ways how to proceed and analyze data, but hey – I'm learning! Sometimes it is better to learn the hard way; explore the language and make mistakes to realize and understand how the languague works behind the scenes.

Before I move to the actuall coding, let's see what data I'm analyzing.

The aim of this project is to analyze two datasets. **CRM system Beamery** and **ATS Jobvite** that are used in Kiwi.com for hiring purposes. Beamery is a tool that Kiwi.com uses for pre-sourcing candindates, for community events applications, and vistors. Jobvite, another HR tool, is used for tracking the applicants, e.g., people who actually applied for an open possition at Kiwi.com. All the data are focused on the engineering department. For GDPR purposes some scripts are not printed because the results contain name or email addresses. **Note that the data are not from the whole company, nor complete.**

Please note that Kiwi.com is using Beamery since May 2019. There are several event(s) data missing due to the lack of proper archivation and (mostly) using spreadsheets that are long gone now. Nontheless, the data analysis can provide us with a glimpse of how the events are performing.

The tables bellow illustrates *some* of the columns from the datasets and their meaning.

___

### Beamery
| Column Name   | Type       | Meaning|
| :-------------|:------------| :----- |
| Name          | string     | Name of the candidate |
| Primary Email | string     | Email of the candidate |
| Source        | string     | Source of the candidate |
| Status        | string     | Are their willing to be contacted by HR? Can be *passive* or *lead* |
| Event         | string     | What event(s) the candidate went on |
| Year          | int        | What year was the event |
| Custom        | string     | What month was the event |

### Jobvite
| Column Name          | Type       | Meaning|
| :-------------       |:------------| :----- |
| Submitted_Date       | datetime     | Name of the candidate |
| Candidate_Full_Name  | string     | Email of the candidate |
| Candidate_Email      | string     | Source of the candidate |
| Candidate_Source_Type| string     | Eg. (1) Job Board or (2) Events |
| Candidate_Source     | string     | Eg. (1) Indeed or (2) Python Weekend Prague June 2019|
| New                  | datetime       | Date when the candidate has submitted the application |
| Screened             | datetime       | Date when the candidate was screended by a recruiter |
| Phone_Screen         | datetime       | Date when the phone screen was done with the candidate |
| Practical_Task       | datetime       | Date when the practical task was send to the candidate |
| Reference_Check      | datetime       | Date when the refference check was done by a recruiter |
| Submitted_to_Hiring_Manager | datetime       | Date when the candidate was submitted to the hiring manager |
| 1st_Interview        | datetime       | Date when the candidate was on the 1st interview |
| 2nd_Interview        | datetime       | Date when the candidate was on the 2nd interview |
| Offer_Sent           | datetime       | Date when the offer was sent to the candidate |
| Hired                | datetime       | Date when the candidate was hired |
| Offer_Rejected       | datetime       | Date when the candidate rejected the offer |
| Rejected             | datetime       | Date when the candidate was rejected by Kiwi.com |
| Candidate_withdrew   | datetime       | Date when the candidate has withdrawn their application |
| Candidate_Disposition| string         | The reason why the candidate has withdrawn their application or was rejected by Kiwi.com |
| Requisition_Id       | int            | ID of the possition the candidate has applied to |
| Job_Title            | datetime       | Title of the possition |
| Requisition_Category | datetime       | Department the candidate has applied to - *in our case Engineering* |

---
### Project KPIs
+ **What is the direct and indirect candidate conversion ( in %)**

   Direct conversion – candidates have a correct Candidate Source Type in Jobvite. This means that it is corresponding with the Beamery dataset. Advanced candidate tracking helps Kiwi.com alocate money better. In addition, we can measure the success rate of events as a hiring source.


+ **What is the quality of the event candidates? ( in %)**
   
   In a discussion we led with the recruitment team we learned that a quality candidate is someone who successfully completes the phone screening. If candidates from events are marked as quality candidates, the workload of the recruitment team decreases and they can source and hire more efficiently.
   
  
+ **How many events precede an actual aplication for an open position?**

   What is the number of events an average candidate goes to before appling? If we know there is more than one event in average, we can prepare better and find patterns.
   

#### What else do we want to know?
+ Ratio of *lead* and *passive* candidates

In [1]:
import csv
from csv import reader


In [2]:
#opening beamery csv as beamery (data) and beamery_header (header row)

opened_file = open("/Users/pavlinaschuster/Documents/DA/Projekt/Data/Beamery_CSV.csv", )
readed_file = reader(opened_file, delimiter=';')
beamery = list(readed_file)
beamery_header = beamery[0]
beamery = beamery[1:]

In [3]:
#opening jobvite csv as jobvite (data) and jobvite_header (header row)

opened_file = open("/Users/pavlinaschuster/Documents/DA/Projekt/Data/Jobvite.csv",)
readed_file = reader(opened_file,  delimiter=';')
jobvite = list(readed_file)
jobvite_header = jobvite[0]
jobvite = jobvite[1:]

In [4]:
#defining function "data_explore" for exploring datasets

def data_explore (dataset, start, end, rows_and_columns=False,):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print("\n") #adding new empty line after each row for better readablity

    if rows_and_columns:
        print("Number of rows:", len(dataset))
        print("Number of columns:", len(dataset[0]))

#### Exploring Beamery dataset, printing header, firsts 3 rows, number of rows, and columns

In [5]:
print(beamery_header)
print("\n")
#data_explore(beamery, 0, 3, True) NOT PRINTED - GDPR

['\ufeffSource.Name', 'Name', 'Primary Photo', 'Company', 'Title', 'Last Activity', 'Last Contact', 'Last Heard From', 'Primary Email', 'Primary Phone Number', 'Status', 'Days in Status', 'Source', 'Location', 'Date Created', 'Interests', 'Keywords', 'Pools', 'Links', 'Vacancies', 'Vacancy Stages', 'Vacancy Date Added', 'Attachments', 'Degree', 'University', 'Contact Profile URL', 'Do Not Contact', 'Unsubscribed Date', 'All Emails', 'All Phone Numbers', 'Global Tags', 'GDPR Consent', 'GDPR Last Consent Date', 'Assigned To', 'Alumni (Ex Employee)', 'Are you a student?', 'Conference Status Attend', 'Conference Status Speak', 'CT e-mail', 'Do you need accommodation?', 'Employee Email', 'Events OK', 'Extra question', 'Helping Community External Talk', 'Helping Community Internal Talk', 'Helping Community Organize IT Event', 'Helping Community Technical Assignments', 'Helping Community Writing', 'How did you learn about the event?', 'HR OK', 'Internal Employee', 'Internal employee.1', 'News

#### Exploring Jobvite dataset, printing header, first 3 rows, number of rows, and columns


In [6]:
print(jobvite_header)
print("\n")
# data_explore(jobvite, 0, 3, True) NOT PRINTED - GDPR

['\ufeffCandidate_Submitted_Date', 'Candidate_Full_Name', 'Candidate_City', 'Candidate_State', 'Candidate_Email', 'Candidate_Source_Type', 'Candidate_Source', 'New', 'Screened', 'Phone_Screen', 'Practical_Task', 'Reference_Check', 'Submitted_to_Hiring_Manager', '1st_Interview', '2nd_Interview', 'Offer_Sent', 'Hired', 'Offer_Rejected', 'Rejected', 'Candidate_withdrew', 'Candidate_Disposition', 'Requisition_Id', 'Job_Title', 'Requisition_Category']




The engineering department at Kiwi.com received 8216 job applications. In Beamery Dataset we have 1970 rows. 

Let's create a list of unique emails in Jobvite now. This list will tell us how many unique people applied for an open position at Kiwi.com. This is not an error, since one person can apply for one or more positions.

In [7]:
jobvite_unique_mail = []
jobvite_duplicate_mail = []

for mails in jobvite:
    mail = mails[4]
    if mail in jobvite_unique_mail:
        jobvite_duplicate_mail.append(mail)
    else:
        jobvite_unique_mail.append(mail)

print("Number of unique mails in Jobvite:", len(jobvite_unique_mail))
print("Number of duplicate mails in Jobvite:", len(jobvite_duplicate_mail))

Number of unique mails in Jobvite: 6534
Number of duplicate mails in Jobvite: 1682


The Jobvite dataset comprises of 6534 unique job applicants.

Let's do the same for Beamery. The number of duplicites will inform us how many people attended more than one event.

In [8]:
beamery_unique_mail = []
beamery_duplicate_mail = []

for mails in beamery:
    mail = mails[8]
    if mail in beamery_unique_mail:
        beamery_duplicate_mail.append(mail)
    else:
        beamery_unique_mail.append(mail)

print("Number of unique mails in Beamery:", len(beamery_unique_mail))
print("Number of duplicate mails in Beamery:", len(beamery_duplicate_mail))

Number of unique mails in Beamery: 1821
Number of duplicate mails in Beamery: 149


Now that we know how many emails we have, let's search for empty values in the email column and assign the non-empty rows back to the Jobvite dataset.

In [9]:
#missing values in Jobvite mails - creating two lists

jv_empty_mails = []
jv_correct_mails = []

for row in jobvite:
    email = row[4]
    if not email:
        jv_empty_mails.append(row)
        # print(row[1]) printing names of ppl who do not have email
    else:
        jv_correct_mails.append(row)


print("Number of empty mails in Jobvite:", len(jv_empty_mails))
print("Number of correct mails in Jobvite:", len(jv_correct_mails))

#assigning the non-empty rows to jobvite dataset
jobvite_correct = jv_correct_mails


Number of empty mails in Jobvite: 18
Number of correct mails in Jobvite: 8198


We have succesfully updated the Jobvite list and deleted the empty emails from the dataset. Now, let's do the same with the Beamery list.

In [10]:
bm_empty_mails = []
bm_correct_mails = []

for row in beamery:
    email = row[8]
    if not email:
        bm_empty_mails.append(row)
        #print(row[1]) printing names of ppl who do not have email
    else:
        bm_correct_mails.append(row)  

print("Number of empty mails in Beamery:", len(bm_empty_mails))
print("Number of correct mails in Beamery:", len(bm_correct_mails))

#assigning the non-empty rows to jobvite dataset
beamery_correct = bm_correct_mails
print("Updated list lenght:",len(beamery))

Number of empty mails in Beamery: 0
Number of correct mails in Beamery: 1970
Updated list lenght: 1970


### What is the direct and indirect candidate conversion (in %)

Now that we have clean datasets of empty emails, we can have a look at our KPIs. We want to learn what the direct and indirect candidate conversion is (in %). To do this, we have to know how many applicantions in Jobvite have source events. Note that the number of applications does not equal the number of people who applied, as one person can have more than one application. There is a second script that tells us the number of people who applied.

In [11]:
#How many applications there is
jobvite_events = []

for row in jobvite_correct:
    source = row[5]
    if source == "Events":
        jobvite_events.append(row)
print("Number of applicantions in Jobvite with source type Events:", len(jobvite_events))



#How many people applied for a position with the source type Events

jobvite_events_unique = []

for row in jobvite_events:
    dis_mail = row[4]
    if dis_mail not in jobvite_events_unique:
        jobvite_events_unique.append(dis_mail)

print("Number of single persons who applied for position with source type Events:", len(jobvite_events_unique))


Number of applicantions in Jobvite with source type Events: 87
Number of single persons who applied for position with source type Events: 83


We can see that the numbers are almost identical; there is not a lot of people from the source type events who would apply for more than one open position. These numbers are based on the source type filled out in Jobvite. However, the data may differ since not all people are properly tracked in the ATS Jobvite. This means that at times the source might not be accurately tagged. The CRM system Beamery is a more relevant source of data. Comparing emails between Beamery and Jobvite will tell us how many applications we actually have from the source type *Events*.

In [12]:
matched_mails = []

for row in beamery_correct:
    bm_mail = row[8]
    for row in jobvite_correct:
        jv_mail = row[4]   
        if bm_mail == jv_mail and not (bm_mail in matched_mails):
            matched_mails.append(bm_mail)

print("Matched mails between Beamery and Jobvite:", len(matched_mails))

wrong_source = len(matched_mails) - len(jobvite_events_unique)
correct_source = len(matched_mails) - wrong_source

#Percentage of corectly tagged ppl
print("Direct conversion:", round(( (correct_source / len(matched_mails)) * 100), 2),"%")

#Percentage of wrongly tagged ppl
print("Indirect conversion:", round(( (wrong_source / len(matched_mails)) * 100), 2),"%")

#List of lists people from events
events = []
events_emails = []

for row in jobvite_correct:
    mail = row[4]
    if not(mail in events_emails) and mail in matched_mails:
        events.append(row)
        events_emails.append(mail)


Matched mails between Beamery and Jobvite: 117
Direct conversion: 70.94 %
Indirect conversion: 29.06 %


#### The direct conversion is 70,94% 
#### The indirect conversion is 29.06%

This numbers above show us that, in fact, **events are almost up to 30% more successful** source of candidates than it was originally thought.

---

### How many events precede an actual aplication for an open position?

What is the average events attendance; how many times at most has a person visited community events?

In [13]:
event_count = {} #empty dictionary, key is mail, value is count of events

for row in beamery_correct:
    bm_mail = row[8]
    if bm_mail not in event_count:
        event_count[bm_mail]=1
    else:
        event_count[bm_mail]+=1

summary = 0 
for value in event_count.values():
    summary += value

mean = round((summary / len(event_count)),0)

max_value = max(event_count.values())

print("Mean:",mean)
print("Max value:", max_value)

Mean: 1.0
Max value: 4


On average, people attend one event before applying for an open position. Maximum of visited events is 4 events per person.

---

### What is the quality of the event candidates? (in %)


In the following section, I will measure the quality of the candidates. As mentioned above, the quality candidate is the one who succesfully finished *Phone Screen* and moved to another stage, excluding the following: *Rejected*, *Offer Rejected* and *Candidate withdrew*.

For the purposes of the analysis, I have created stages that will be used in the following dictionary. The stages are:

| Status in the recruitment workflow | Stage   | Index no   |
| :------------------------------|:--------| :--------- |
| New                            | Stage 1 | index no 7 |
| Screened                       | Stage 2 | index no 8 |
| Phone_Screen                   | Stage 3 | index no 9 |
| Practical_Task                 | Stage 4 | index no 10|
| Reference_Check                | Stage 5 | index no 11|
| Submitted_to_Hiring_Manager    | Stage 6 | index no 12|
| 1st_Interview                  | Stage 7 | index no 13|
| 2nd_Interview                  | Stage 8 | index no 14|
| Offer_Sent                     | Stage 9 | index no 15|
| Hired                          | Stage 10| index no 16|


In [14]:
success_dict = {}

for row in events:
    mail = row[4]
    source = row[5]
    new = 0 if row[7] is "" else 1
    screened = 0 if row[8] is "" else 1
    phone_screen = 0 if row[9] is "" else 1
    practical_task = 0 if row[10] is "" else 1
    reference_check = 0 if row[11] is "" else 1
    hiring_manager = 0 if row[12] is "" else 1
    first_interview = 0 if row[13] is "" else 1
    second_interview = 0 if row[14] is "" else 1
    offer_sent = 0 if row[15] is "" else 1
    hired = 0 if row[16] is "" else 1

    stages = [hired, offer_sent, second_interview, first_interview, hiring_manager, reference_check, practical_task, phone_screen, screened, new]
    for i in range(len(stages)):
        if stages[i] == 1:
            success_dict[mail] = len(stages) - i
            break


The previous script created a dictionary with emails as keys and stages as values. Now we need to learn how many of those are quality candidates. Quality candidate is anyone above the Stage 3.

In [15]:
def keys_by_values(some_dictionary, list_of_values):
    list_of_keys = list()
    list_of_items = some_dictionary.items()
    for item  in list_of_items:
        if item[1] in list_of_values:
            list_of_keys.append(item[0])
    return  list_of_keys 


list_of_keys = keys_by_values(success_dict, [4,5,6,7,8,9,10,] )
print("Percentage of quality candidates:", round(( len(list_of_keys) / len(success_dict) * 100 )), "%" )

Percentage of quality candidates: 48 %


#### There is 48% success rate of quality candidates from the events source type.


### Number of lead versus passive event participants

The next question we have to ask is how many people who have visited an event have current status *lead* or *passive*. Those two statuses change after recruiters performs a candidate screening. The lead and passive status tell us how many people want to be contacted right now (lead) and how many people want to be contacted in the future (passive). The number of leads stands for the current recruitment workload that needs to be processed. The number of passive candidates represents how many potentital candidates Kiwi.com has for a specific role that is going to be opened in the future.

In [16]:
#number of leads and passives

lead =[]
passive = []
other = []

for row in beamery_correct:
    status = row[10]
    if status == "Lead":
        lead.append(status)
    elif status == "Passive":
        passive.append(status)
    else:
        other.append(status)

print("Number of leads:", len(lead))
print("Number of passives:", len(passive))
print("Number of other statuses:", len(other))

Number of leads: 283
Number of passives: 446
Number of other statuses: 1241


**In total, there is 729 people who are interested in working at Kiwi.com** and have not yet been contacted by a recruiter.
